src/compiler-debugging.md

```diff
@@ -0,0 +1,320 @@
+needed!
+
+This means that you should not put anything too expensive or likely
+to crash there - that would annoy anyone who wants to use logging for their own 
+module. Note that if `RUST_LOG` is unset (the default), then the code will not 
+run - this means that if your logging code panics, then no-one will know it 
+until someone tries to use logging to find *another* bug.
+
+If you *need* to do an expensive operation in a log, be aware that while log 
+expressions are *evaluated* even if logging is not enabled in your module, 
+they are not *formatted* unless it *is*. This means you can put your 
+expensive/crashy operations inside an `fmt::Debug` impl, and they will not be 
+run unless your log is enabled:
+
+```Rust
+use std::fmt;
+
+struct ExpensiveOperationContainer<'a, 'gcx, 'tcx>
+    where 'tcx: 'gcx, 'a: 'tcx
+{
+    tcx: TyCtxt<'a, 'gcx, 'tcx>
+}
+
+impl<'a, 'gcx, 'tcx> fmt::Debug for ExpensiveOperationContainer<'a, 'gcx, 'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let value = random_operation(tcx);
+        fmt::Debug::fmt(&value, fmt)
+    }
+}
+
+debug!("{:?}", ExpensiveOperationContainer { tcx });
+```
+
+## Formatting Graphviz output (.dot files)
+[formatting-graphviz-output]: #formatting-graphviz-output
+
+Some compiler options for debugging specific features yield graphviz graphs - 
+e.g. the `#[rustc_mir(borrowck_graphviz_postflow="suffix.dot")]` attribute
+dumps various borrow-checker dataflow graphs.
+
+These all produce `.dot` files. To view these files, install graphviz (e.g.
+`apt-get install graphviz`) and then run the following commands:
+
+```
+$ dot -T pdf maybe_init_suffix.dot > maybe_init_suffix.pdf
+$ firefox maybe_init_suffix.pdf # Or your favorite pdf viewer
+```
+
+## Debugging LLVM
+[debugging-llvm]: #debugging-llvm
+
+LLVM is a big project on its own that probably needs to have its own debugging
+document (not that I could find one). But here are some tips that are important
+in a rustc context:
+
+The official compilers (including nightlies) have LLVM assertions disabled,
+which means that LLVM assertion failures can show up as compiler crashes (not
+ICEs but "real" crashes) and other sorts of weird behavior. If you are
+encountering these, it is a good idea to try using a compiler with LLVM
+assertions enabled - either an "alt" nightly or a compiler you build yourself
+by setting `[llvm] assertions=true` in your config.toml - and
+see whether anything turns up.
+
+The rustc build process builds the LLVM tools into 
+`./build/<host-triple>/llvm/bin`. They can be called directly.
+
+The default rustc compilation pipeline has multiple codegen units, which is hard
+to replicate manually and means that LLVM is called multiple times in parallel.
+If you can get away with it (i.e. if it doesn't make your bug disappear),
+passing `-C codegen-units=1` to rustc will make debugging easier.
+
+If you want to play with the optimization pipeline, you can use the `opt` from
```

Comment:
Perhaps rephrase this sentence as "If you want to play with the optimization pipeline, you can use the `opt` tool from `./build/<host-triple>/llvm/bin/` with the the LLVM IR emitted by rustc."

Reply:
Changed the first sentence and ended up with this paragraph! 

>If you want to play with the optimization pipeline, you can use the opt tool 
from `./build/<host-triple>/llvm/bin/` with the the LLVM IR emitted by rustc. 
Note that rustc emits different IR depending on whether `-O` is enabled, even 
without LLVM's optimizations, so if you want to play with the IR rustc emits, 
you should:

let me know if it's fine!